src/encoder.ts

```diff
@@ -14,119 +14,229 @@
-  append(
-    buf,
-    Uint8Array.from([Kind.Array, valueKind]),
-    encodeUint32(new Uint8Array(), size)
-  );
-
-export const encodeMap = (
-  buf: Uint8Array,
-  size: number,
-  keyKind: Kind,
-  valueKind: Kind
-) =>
-  append(
-    buf,
-    Uint8Array.from([Kind.Map, keyKind, valueKind]),
-    encodeUint32(new Uint8Array(), size)
-  );
-
-export const encodeUint8Array = (buf: Uint8Array, value: Uint8Array) =>
-  append(
-    buf,
-    Uint8Array.from([Kind.Uint8Array]),
-    encodeUint32(new Uint8Array(), value.length),
-    value
-  );
-
-export const encodeString = (buf: Uint8Array, value: string) => {
-  const v = new TextEncoder().encode(value);
-
-  return append(
-    buf,
-    Uint8Array.from([Kind.String]),
-    encodeUint32(new Uint8Array(), v.length),
-    v
-  );
-};
-
-export const encodeError = (buf: Uint8Array, value: Error) => {
-  const v = new TextEncoder().encode(value.message);
-
-  return append(
-    buf,
-    Uint8Array.from([Kind.Error, Kind.String]),
-    encodeUint32(new Uint8Array(), v.length),
-    v
-  );
-};
+const CONTINUATION = 0x80;
+const PRE_CONTINUATION = 0x7f;
+const REST_BYTES = 7;
+
+export class Encoder {
+  #pos = 0;
+
+  #buf: Uint8Array;
+
+  constructor(buf?: Uint8Array) {
+    if (buf === undefined) {
+      this.#buf = new Uint8Array(512);
+    } else {
+      this.#buf = buf;
+    }
+  }
+
+  private resize(minLen: number) {
+    if (this.#buf.length - this.#pos < minLen) {
+      const oldBuf = this.#buf;
+
+      this.#buf = new Uint8Array(this.#buf.buffer.byteLength * 2);
+      this.#buf.set(oldBuf);
+    }
+  }
+
+  get bytes() {
+    return new Uint8Array(this.#buf.buffer, 0, this.#pos);
+  }
+
+  null() {
+    this.resize(1);
+    this.#buf[this.#pos] = Kind.Null;
+    this.#pos += 1;
+    return this;
+  }
+
+  any() {
+    this.resize(1);
+    this.#buf[this.#pos] = Kind.Any;
+    this.#pos += 1;
+    return this;
+  }
+
+  boolean(value: boolean) {
+    this.resize(2);
+    this.#buf[this.#pos] = Kind.Boolean;
+    this.#buf[this.#pos + 1] = value ? BOOLEAN_TRUE : BOOLEAN_FALSE;
+    this.#pos += 2;
+    return this;
+  }
+
+  uint8(value: number) {
+    this.resize(2);
+    this.#buf[this.#pos] = Kind.Uint8;
+
+    const dataView = new DataView(Uint8Array.from([0]).buffer);
+    dataView.setUint8(0, value);
+    const bytes = new Uint8Array(dataView.buffer);
+
+    this.#buf.set(bytes, this.#pos + 1);
+    this.#pos += bytes.length + 1;
+    return this;
+  }
+
+  uint16(value: number) {
+    let val = value;
+    this.resize(4);
+    this.#buf[this.#pos] = Kind.Uint16;
+    this.#pos += 1;
+
+    while (val >= CONTINUATION) {
+      this.#buf[this.#pos++] = val | CONTINUATION;
+      val >>>= REST_BYTES;
+    }
+    this.#buf[this.#pos++] = val;
+    return this;
+  }
+
+  uint32(value: number) {
+    let val = value;
+    this.resize(6);
+    this.#buf[this.#pos] = Kind.Uint32;
+    this.#pos += 1;
+
+    while (val >= CONTINUATION) {
```

Comment:
Same here and below, probably a good idea to merge the bitwise operation/loop into a private func